src/rgw/rgw-restore-bucket-index

```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env bash
+
+# version 2023-01-11
+
+# IMPORTANT: affects order produced by 'sort' and 'ceph-diff-sorted'
+# relies on this ordering
+export LC_ALL=C
+
+bkt_entry=/tmp/bkt-entry.$$
+bkt_inst=/tmp/bkt-inst.$$
+obj_list=/tmp/object-list.$$
+export data_pools=/tmp/data-pools.$$
+
+log() {
+  echo $(date +%F\ %T) $(hostname -s) "$1"
+}
+
+usage() {
+  >&2 cat << EOF
+
+Usage: $0 <bucket-name> [data-pool-name]
+EOF
+  exit 1
+}
+
+get_field() {
+  grep "\"$2\":" $1 | sed -E "s/.*\".*\": ([^,]*),?/\1/"
+}
+
+# quotes
+get_field_q() {
+  grep "\"$2\":" $1 | sed -E "s/.*\".*\": \"(.*)\".*/\1/"
+}
+
+get_pool() {
+  if [ -n "$1" ] ;then
+    echo "$1"
+  else
+    radosgw-admin zone get >$data_pools 2>/dev/null
+    pool_count=$(grep "\"data_pool\":" $data_pools | wc -l)
+    if [ "$pool_count" -ne 1 ] ;then
+      echo "Error: unable to determine data pool; please specify on command-line."
+      usage
+    fi
+    get_field_q $data_pools data_pool
+  fi
+}
+
+clean() {
+  # rm -f $bkt_entry $bkt_inst $obj_list $data_pools
+  echo clean
+}
+
+if [ $# -eq 0 -o $# -gt 2 ] ;then
+   usage
+fi
+
+bucket=$1
+
+set -e # exit if get_pool fails
+pool=$(get_pool $2)
+set +e
+
+echo pool is $pool
+
+radosgw-admin metadata get bucket:$bucket >$bkt_entry 2>/dev/null
+
+marker=$(get_field_q $bkt_entry marker)
+bucketid=$(get_field_q $bkt_entry bucket_id)
+
+echo marker is $marker
+echo bucketid is $bucketid
+
+radosgw-admin metadata get bucket.instance:${bucket}:$bucketid >$bkt_inst 2>/dev/null
+
+set -e
+bkt_flags=$(get_field $bkt_inst flags)
+num_shards=$(get_field $bkt_inst num_shards)
+set +e
+
+is_versioned=$(( $bkt_flags & 2))
+if [ "$is_versioned" -ne 0 ] ;then
+   echo "Error: this bucket appears to be versioned, and this tool cannot work with versioned buckets."
+   exit 2
+fi
+
+echo num_shards is $num_shards
+
+if [ "$num_shards" -le 0 ] ;then
+   echo "Error: number of shards -- $num_shards -- is invalid"
+   exit 3
+fi
+
+(rados -p $pool ls | grep "^${marker}_" | sed "s/^${marker}_\(.*\)/\1/" >$obj_list ) 2>/dev/null
```

Comment:
isn't this going to pick up 'namespaced' rados objects like multipart parts and shadow/tail objects?

Reply:
Good catch; need to prohibit the next underscore.